pkg/build/api/types.go

```diff
@@ -352,6 +356,25 @@ type SourceBuildStrategy struct {
 	ForcePull bool
 }
 
+// BuildHookSpec holds build hooks with actions grouped by build status.
+type BuildHookSpec struct {
+	// The OnSuccess group is a list of BuildHooks that will be executed
+	// when the build changes to BuildPhaseComplete.
+	OnSuccess []BuildHook
+
+	// The OnFailure group is a list of BuildHooks that will be executed
+	// when the build changes to BuildPhaseFailed.
+	OnFailure []BuildHook
+}
+
+// A BuildHook defines actions to be executed when a build changes to a certain
+// phase.
+type BuildHook struct {
+	// StartBuilds holds a list of references to BuildConfigs that will be
+	// started when the hook is executed.
+	StartBuilds []kapi.ObjectReference
+}
```

Comment:
I think this whole nesting can be simplified.  Here's one quick thought: Start with BuildSpec having a new field:
PostSteps []BuildStep

type BuildStep struct {
  // this step will be run when the build enters any phase in this array
  // now we can support any phase the user wants, not just success/fail.
  onStatus []BuildPhaseType
  // action this step takes when invoked
  action Action
  // set to true once the action has been successfully processed
  // used by controller to determine which actions still need to be run
  complete bool
}

type Action interface {
  // bool indicates if action should be retried, error indicates the error that occured if any
  func Invoke() bool,error
}

// implements Action
type InvokeBuildAction struct {
  Build kapi.ObjectReference
}